### PR TITLE
`azurerm_app_configuration` - retry `checkNameAvailability`  to make sure delete is finished, add stateWait to confirm key/feature is created

### PR DESCRIPTION
**Pull request by teowa**
resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/21939
- add the `state.WaitForStateContext` to fix `Root resource was present, but now absent` error of `azurerm_app_configuration_key/feature`.
- add the `state.WaitForStateContext` to retry `checkNameAvailability` accoding to https://github.com/hashicorp/terraform-provider-azurerm/pull/17714#issuecomment-1225449556.
```
TF_ACC=1 go test -v ./internal/services/appconfiguration -parallel 50 -test.run=TestAccAppConfiguration_softDeletePurgeThenRecreate -timeout 1440m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAppConfiguration_softDeletePurgeThenRecreate
=== PAUSE TestAccAppConfiguration_softDeletePurgeThenRecreate
=== CONT  TestAccAppConfiguration_softDeletePurgeThenRecreate
--- PASS: TestAccAppConfiguration_softDeletePurgeThenRecreate (556.89s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appconfiguration      558.142s

```

### CODE REVIEW

**Review by stephybun — CHANGES_REQUESTED** [review]
```
------- Stdout: -------
=== RUN   TestAccAppConfiguration_softDeleteRecovery
=== PAUSE TestAccAppConfiguration_softDeleteRecovery
=== CONT  TestAccAppConfiguration_softDeleteRecovery
testcase.go:110: Step 4/5 error: Error running apply: exit status 1
Error: waiting for the Configuration Store Configuration Store (Subscription: "*******"
Resource Group Name: "acctestRG-appconfig-230511054705781851"
Configuration Store Name: "testaccappconf230511054705781851") Name Available: timeout while waiting for state to become 'Available' (last state: 'Unavailable', timeout: 29m59.517047832s)
with azurerm_app_configuration.test,
on terraform_plugin_test.tf line 22, in resource "azurerm_app_configuration" "test":
22: resource "azurerm_app_configuration" "test" {
--- FAIL: TestAccAppConfiguration_softDeleteRecovery (1892.83s)
FAIL
```

**Review by katbyte — CHANGES_REQUESTED** [review]
I'm getting a test failure:
```
------- Stdout: -------
=== RUN   TestAccAppConfigurationKey_basicVault
=== PAUSE TestAccAppConfigurationKey_basicVault
=== CONT  TestAccAppConfigurationKey_basicVault
    testcase.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: Provider produced inconsistent result after apply
        
        When applying changes to azurerm_app_configuration_key.test, provider
        "provider[\"registry.terraform.io/hashicorp/azurerm\"]" produced an
        unexpected new value: Root resource was present, but now absent.
        
        This is a bug in the provider, which should be reported in the provider's own
        issue tracker.
--- FAIL: TestAccAppConfigurationKey_basicVault (409.06s)
FAIL
```

**Comment by katbyte** [comment]
still seeing test failures
image

**Comment by teowa** [comment]
Need to fix tests

**Comment by teowa** [comment]
image

File: internal/services/appconfiguration/app_configuration.go [hunk 1]
```diff
@@ -30,6 +30,9 @@ func appConfigurationGetKeyRefreshFunc(ctx context.Context, client *appconfigura
 				if utils.ResponseWasForbidden(autorest.Response{Response: v.Response}) {
 					return "Forbidden", "Forbidden", nil
 				}
+				if utils.ResponseWasNotFound(autorest.Response{Response: v.Response}) {
```
Comment (stephybun) [inline internal/services/appconfiguration/app_configuration.go:33]: Can we make this change above as well for `response.WasForbidden`
```suggestion
				if response.WasNotFound(v.Response) {
```

File: internal/services/appconfiguration/app_configuration_feature_resource.go [hunk 2]
```diff
@@ -212,17 +212,18 @@ func (k FeatureResource) Create() sdk.ResourceFunc {
 
 			// from https://learn.microsoft.com/en-us/azure/azure-app-configuration/concept-enable-rbac#azure-built-in-roles-for-azure-app-configuration
 			// allow some time for role permission to be done propagated
-			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Key %q read permission to be done propagated", featureKey)
+			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Feature %q read permission to be done propagated", featureKey)
```
Comment (stephybun) [inline internal/services/appconfiguration/app_configuration_feature_resource.go:215]:
```suggestion
			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Feature %q read permission to be propagated", featureKey)
```

File: internal/services/appconfiguration/app_configuration_feature_resource.go [hunk 3]
```diff
@@ -243,6 +244,20 @@ func (k FeatureResource) Create() sdk.ResourceFunc {
 				return fmt.Errorf("while creating feature: %+v", err)
 			}
 
+			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Feature %q to be done provisioned", model.Key)
```
Comment (stephybun) [inline internal/services/appconfiguration/app_configuration_feature_resource.go:247]:
```suggestion
			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Feature %q to be provisioned", model.Key)
```

File: internal/services/appconfiguration/app_configuration_key_resource.go [hunk 4]
```diff
@@ -207,6 +208,20 @@ func (k KeyResource) Create() sdk.ResourceFunc {
 				}
 			}
 
+			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Key %q to be done provisioned", model.Key)
```
Comment (stephybun) [inline internal/services/appconfiguration/app_configuration_key_resource.go:211]:
```suggestion
			metadata.Logger.Infof("[DEBUG] Waiting for App Configuration Key %q to be provisioned", model.Key)
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Can you please add a test for `azurerm_app_configuration_feature` and `azurerm_app_configuration_key` to validate that the wait functions after creation is actually needed?

**Comment by teowa** [comment]
Hi @stephybun, thanks for reviewing this. I think it is hard to validate wait function is needed, for the `produced an unexpected new value: Root resource was present, but now absent` error is not 100% reproducible. The wait function is added based on the error from test subscription and customer provided error message in https://github.com/hashicorp/terraform-provider-azurerm/issues/21939. I have also report the error to service team through https://github.com/Azure/AppConfiguration/issues/733

**Pull request merged**